### PR TITLE
Add Period Start and End functions to Date and DateTime

### PR DESCRIPTION
**Pull request by radeusgd**
### Pull Request Description

Implements https://www.pivotaltracker.com/story/show/183081152

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

File: std-bits/base/src/main/java/org/enso/base/time/Date_Period_Utils.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+package org.enso.base.time;
+
+import java.time.temporal.ChronoField;
+import java.time.temporal.Temporal;
+import java.time.temporal.TemporalAdjusters;
+
+public class Date_Period_Utils implements TimeUtilsBase {
```
Comment (radeusgd) [inline std-bits/base/src/main/java/org/enso/base/time/Date_Period_Utils.java:6]: @jdunkerley do you think I should add some doc comments? Seems a bit 'naked' without them but OTOH I'd say the name of the function seems to be pretty self explanatory so I'm not sure if the comments would add any clarity. What do you think?

**Review by 4e6 — APPROVED** [review]
Thanks for updating the SBT build formatting

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date.enso [hunk 2]
```diff
@@ -245,7 +254,7 @@ type Date
 
              example_to_time = Date.new 2020 2 3 . to_time Time_Of_Day.new Time_Zone.utc
     to_time : Time_Of_Day -> Time_Zone -> Date_Time
-    to_time self time_of_day (zone=Time_Zone.system) = self.to_time_builtin time_of_day zone
+    to_time self (time_of_day=Time_Of_Day.new) (zone=Time_Zone.system) = self.to_time_builtin time_of_day zone
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date.enso]: This should be called `to_date_time` now I think.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Time_Of_Day.enso [hunk 3]
```diff
@@ -193,7 +202,7 @@ type Time_Of_Day
              from Standard.Base import Time_Of_Day
 
              example_to_time = Time_Of_Day.new 12 30 . to_time (Date.new 2020)
-    to_time : Date -> Time_Zone -> Time
+    to_time : Date -> Time_Zone -> Date_Time
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Time_Of_Day.enso]: as above `to_date_time`

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date_Time.enso [hunk 4]
```diff
@@ -326,6 +328,24 @@ type Date_Time
     day_of_week self =
         Day_Of_Week.from (Time_Utils.get_field_as_zoneddatetime self ChronoField.DAY_OF_WEEK) Day_Of_Week.Monday
 
+    ## Returns the first date within the `Time_Period` or `Date_Period`
+       containing self.
+    start_of : (Date_Period|Time_Period) -> Date_Time
+    start_of self period=Date_Period.Month =
+        adjusted = period.adjust_start self
+        case period.is_date_period of
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date_Time.enso:336]: Can we use `is_a Date_Period`?

File: std-bits/base/src/main/java/org/enso/base/time/Date_Period_Utils.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
```
Comment (jdunkerley) [inline std-bits/base/src/main/java/org/enso/base/time/Date_Period_Utils.java:6]: We could do this as:
```
public class FirstDayOfQuarter implements TemporalAdjuster {
  @Override
  public Temporal adjustInto(Temporal temporal) {
    int currentQuarter = YearMonth.from(temporal).get(IsoFields.QUARTER_OF_YEAR);
    return temporal.with(IsoFields.QUARTER_OF_YEAR, currentQuarter).with(TemporalAdjusters.firstDayOfMonth());
  }
}
```

and just have the single adjust flow?

**Review by jdunkerley — APPROVED** [review]
One minor rename and a suggestion.
But looks good.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date_Time.enso (hunk 4 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date_Time.enso:336]: Screenshot from 2022-09-10 14-00-32

Unfortunately not. Until we rework `is_a` to work with types.

File: std-bits/base/src/main/java/org/enso/base/time/Date_Period_Utils.java (hunk 5 above)
Comment (radeusgd) [reply std-bits/base/src/main/java/org/enso/base/time/Date_Period_Utils.java:6]: That's a splendid idea! I will try to implement it.

**Comment by radeusgd** [comment]
I could not fully test the autumn DST switch case as the overlapped timelines are not handled properly in polyglot which makes our Enso Date_Time unable to express one of the two instants in time there (2:30 am _after_ the DST switch is not expressible, it gets converted back to 2:30 am _before_ switch which is an hour earlier).

Reported this at https://github.com/oracle/graal/issues/4918

**Review by hubertp — APPROVED** [review]

**Pull request merged**